### PR TITLE
Add sinc operator

### PR DESCRIPTION
**Pull request by soulitzer**
Implements the sinc operator. 
See https://numpy.org/doc/stable/reference/generated/numpy.sinc.html

image

### CODE REVIEW

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -302,6 +302,17 @@ static void sgn_kernel(TensorIterator& iter){
   });
 }
 
+static void sinc_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(kBFloat16, iter.common_dtype(), "sinc_cpu", [&]() {
+    cpu_kernel(
+        iter,
+        [=](scalar_t a) -> scalar_t {
+          scalar_t product = scalar_t(M_PI) * a;
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:310]: Do the check for the zero special-case before computing the product and add a comment explaining why it's special-cased.

File: aten/src/ATen/native/cuda/UnaryOpsKernel.cu [hunk 2]
```diff
@@ -103,6 +103,15 @@ void sigmoid_kernel_cuda(TensorIterator& iter) {
   });
 }
 
+void sinc_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.common_dtype(), "sinc_cuda", [&]() {
+    gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
+      scalar_t product = scalar_t(M_PI) * a;
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/UnaryOpsKernel.cu:109]: Same as with CPU kernel: perform the check for zero before computing the product (since the product computation is not necessary if the input is zero)

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -7453,6 +7453,34 @@ def merge_dicts(*dicts):
     tensor([-0.5194,  0.1343, -0.4032, -0.2711])
 """.format(**common_args))
 
+add_docstr(torch.sinc,
+           r"""
+sinc(input, *, out=None) -> Tensor
+
+Returns a new tensor with the normalized sinc of the elements of :attr:`input`.
```
Comment (mruberry) [inline torch/_torch_docs.py:7460]: "Returns a new tensor with the normalized sinc..." -> "Computes the normalized sinc of :attr:\`input\.`

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -281,6 +281,15 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
                                     low=None, high=None,
                                     requires_grad=False))),)
 
+def np_sinc_with_fp16_as_fp32(x):
+    # Wraps numpy's sinc function so that fp16 values are promoted to fp32
+    # before sinc is invoked. Context: numpy's sinc returns NaN when evaluated
+    # at 0 for fp16.
+    if x.dtype in [np.float16]:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:288]: -> `if x.dtype is np.float16:`

**Review by mruberry — APPROVED** [review]
Nice work, @soulitzer! Just some minor nits for cleanup before merging. Thanks for being so diligent with adding an OpInfo for `sinc`.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (soulitzer) [reply torch/testing/_internal/common_methods_invocations.py:288]: Testing with `is` fails because of a bug https://stackoverflow.com/questions/26921836/correct-way-to-test-for-numpy-dtype. Using `==` here instead.

**Comment by mruberry** [comment]
The asan_test2 build failure is unrelated to this PR.
